### ISSUE TITLE
How to decode from the dev data?

### ISSUE THREAD

**Issue by appleyib** [mravanelli/pytorch-kaldi#6]
The cfg files specify $data_dir (which is $KALDI_ROOT....../data/test/ in default). For the purpose to get the LMWT that has the lowest WER in dev set(./RESULTS will only use the one that has lowest WER in **test set**), I changed the $data_dir to let it points to dev data folder($KALDI_ROOT/....../data/dev).

However, decoding process failed due to feature/frame files that were being decoded do not match the files with the correct labels among all score 0 - 9. Probably there is still something pointing to the test set.

Is there any guide for decoding from dev data set of KALDI?

**Comment by mravanelli** [comment]
Usually Kaldi recipes decode the test set with different LM weights and take the best performance. In this project, we decided to directly inherit this kaldi approach even if I totally agree that a more fair way to evaluate the performance would be to decode the dev_set and use its best LM weight to decode the  test set. In your case it sounds like you are using the wrong labels (i.e., the wrong alignments). Make sure that te_lab_folder='$KALDI_ROOT/egs/timit_s5_git/exp/dnn4_pretrain-dbn_dnn_ali_dev' (in your case it should contain dev labels and not test labels).

**Comment by appleyib** [comment]
Thank you for you quick reply. After modifying the te_lab_folder and re-running the run_exp.sh,  it still failed due to same reason.  Is deleting all the files in the output folder and complete re-train necessary? Also, should ali_dir be changed to '$KALDI_ROOT/egs/timit_s5_git/exp/dnn4_pretrain-dbn_dnn_ali_dev' as well?

**Comment by mravanelli** [comment]
What is exactly the error you get?

**Comment by appleyib** [comment]
The following is the last line of the prompt:

run.pl: 10 / 10 failed, log is in ../exp/TIMIT_liGRU_fmllr_Third_Try/decoding_dev/scoring/log/score.*.log

The log files in the above line have repeated similar error messages such like:
Performing alignments for file 'mwjg0_sx44'.
1 of 1 Segments For Channel 1. align_ctm_to_stm: Error: Hyp file has more data than ref file, beginning at hyp file/channel 'fdhc0_si1559' '1'.
sclite: Alignment failed.  Exiting.

**Comment by mravanelli** [comment]
You should probably just remove the decoding_dev folder and run the script
again. Does this solve the issues?

**Comment by appleyib** [comment]
Not actually, still same error messages generated. Again, I only modified **$data_dir** and **$te_lab_folder** to let them point to **$KALDI_ROOT/....../data/dev** and **$KALDI_ROOT/egs/timit_s5_git/exp/dnn4_pretrain-dbn_dnn_ali_dev**.

It worked when only modifying **$te_lab_folder** though, but I do not think keeping **$dara_dir** pointing to test set will get me what I want.

**Comment by mravanelli** [comment]
I see, that's weird. In the error you show me before the "* 'mwjg0_sx44'*"
id is part of the dev set (see cat
~/kaldi-trunk/egs/timit/s5/data/dev/wav.scp) while the alignments contain
the file "*fdhc0_si1559*" that is part of the test set (see cat
~/kaldi-trunk/egs/timit/s5/data/test/wav.scp). In practice there is a
feature-label mismatch originated somewhere. I'm surprised that simply
deleting the decode_dev folder and run the decoding again doesn't solve the
issue (sometimes during decoding some steps are skipped and it is usually
better to delete the folder if you change the labels).

Could you confirm that your test part of the [data] section in the cfg file
now looks like that?

te_fea_scp=/home/ravanelm/pytorch_exp/mfcc_dev/dev_split.000
te_fea_opts="apply-cmvn
--utt2spk=ark:/home/ravanelm/kaldi-trunk/egs/timit_s5_git/data/dev/utt2spk
ark:/home/ravanelm/pytorch_exp/mfcc_dev/dev_cmvn_speaker.ark ark:- ark:- |
add-deltas --delta-order=2 ark:- ark:- |"
te_lab_folder='/home/ravanelm/kaldi-trunk/egs/timit_s5_git/exp/dnn4_pretrain-dbn_dnn_ali_dev'
te_lab_opts='ali-to-pdf'

**Comment by appleyib** [comment]
Well I had only changed **$te_lab_folder** and **$data_dir** before, but did change all of those this time. Again, similar error messages got output. The following is the example:

**Performing alignments for file 'mwjg0_sx314'.
        1 of 1 Segments For Channel 1. align_ctm_to_stm: File identifiers do not match but continuing. ref file/channel 'mwjg0_sx404' '1', next hyp 'fdhc0_si1559' '1'.
    
    Performing alignments for file 'mwjg0_sx404'.
        1 of 1 Segments For Channel 1. align_ctm_to_stm: File identifiers do not match but continuing. ref file/channel 'mwjg0_sx44' '1', next hyp 'fdhc0_si1559' '1'.

Performing alignments for file 'mwjg0_sx44'.
1 of 1 Segments For Channel 1. align_ctm_to_stm: Error: Hyp file has more data than ref file, beginning at hyp file/channel 'fdhc0_si1559' '1'.
sclite: Alignment failed.  Exiting.**

Just to make it clear, all the changes that I have made on config file this time are just $te_lab_folder, $te_lab_opts, $te_fea_opts, $te_fea_scp (those were changed to what you mentioned above) and $data_dir (pointing to dev data set).

**Comment by mravanelli** [comment]
yes, what you have done is exactly what one should do to change the
decoding data. Given that, it's totally unclear to me why some alignments
coming from the test still survive even if you delete decode_test (in the
past I never observed something like this ). Note that you have to delete
decoding_test folder and not decoding_dev as I report by mistake in the
previous answer.

**Comment by appleyib** [comment]
There had been only decoding_dev folder in my output folder since I had changed the ./run_exp.sh to rename the decoding folder. I just deleted that folder, made sure that decoding_test was not there and ran run_exp.sh with exactly same config file above. As a result, same error messages were generated. That looked really a bit weird.
Same to my previous idea, I think it is probably due to there is something out of cfg file still pointing to test data sets or labels.

**Comment by mravanelli** [comment]
well, there should be a mess somewhere. I haven't seen errors like this in
past when changing the test set. Could you confirm that with the standard
setup (i.e., without changing the test set) everything is working?

**Comment by mravanelli** [comment]
Also,
just to be sure try also to delete "out_file=$out_folder"/forward_ep_"$epoch
"_ck_"$chunk".pkl"" which is the file containing the posterior
probabilities for each sentence. It could be outdated if you changed the
test set....

**Comment by mravanelli** [comment]
and also...conf_file=$out_folder"/forward_ep_"$epoch"_ck_"$chunk".cfg"

**Comment by mravanelli** [comment]
You are welcome! thank you for the interest in my repository and let me know if you have other issues. We are working on a new code release for the pytorch-kaldi project that will be hopefully ready in a month....

**Comment by appleyib** [comment]
That sounds awesome! Again, thank you for posting this repository!
